strawberry/experimental/pydantic2/utils.py

```diff
@@ -0,0 +1,144 @@
+    get_optional_annotation,
+    is_list,
+    is_optional,
+)
+
+if TYPE_CHECKING:
+    from pydantic import BaseModel
+    from pydantic.fields import ModelField
+    from pydantic.typing import NoArgAnyCallable
+
+
+def normalize_type(type_: Type) -> Any:
+    if is_list(type_):
+        return List[normalize_type(get_list_annotation(type_))]  # type: ignore
+
+    if is_optional(type_):
+        return get_optional_annotation(type_)
+
+    return type_
+
+
+def get_strawberry_type_from_model(type_: Any) -> Any:
+    if hasattr(type_, "_strawberry_type"):
+        return type_._strawberry_type
+    else:
+        raise UnregisteredTypeException(type_)
+
+
+def get_private_fields(cls: Type) -> List[dataclasses.Field]:
+    private_fields: List[dataclasses.Field] = []
+
+    for field in dataclasses.fields(cls):
+        if is_private(field.type):
+            private_fields.append(field)
+
+    return private_fields
+
+
+class DataclassCreationFields(NamedTuple):
+    """Fields required for the fields parameter of make_dataclass"""
+
+    name: str
+    field_type: Type
+    field: dataclasses.Field
+
+    def to_tuple(self) -> Tuple[str, Type, dataclasses.Field]:
+        # fields parameter wants (name, type, Field)
+        return self.name, self.field_type, self.field
+
+
+def is_required(field: ModelField) -> bool:
+    return field.default is PydanticUndefinedType and field.default_factory is None
+
+
+def get_default_factory_for_field(
+    field: ModelField,
+) -> Union[NoArgAnyCallable, dataclasses._MISSING_TYPE]:
+    """
+    Gets the default factory for a pydantic field.
+
+    Handles mutable defaults when making the dataclass by
+    using pydantic's smart_deepcopy
+
+    Returns optionally a NoArgAnyCallable representing a default_factory parameter
+    """
+    # replace dataclasses.MISSING with our own UNSET to make comparisons easier
+    default_factory = (
+        field.default_factory if field.default_factory is not None else UNSET
+    )
+    default = (
+        field.default if not isinstance(field.default, PydanticUndefinedType) else UNSET
+    )
+
+    has_factory = default_factory is not UNSET
+    has_default = default is not UNSET
+
+    # defining both default and default_factory is not supported
+
+    if has_factory and has_default:
+        default_factory = cast("NoArgAnyCallable", default_factory)
+
+        raise BothDefaultAndDefaultFactoryDefinedError(
+            default=default, default_factory=default_factory
+        )
+
+    # if we have a default_factory, we should return it
+
+    if has_factory:
+        default_factory = cast("NoArgAnyCallable", default_factory)
+
+        return default_factory
+
+    # if we have a default, we should return it
+
+    if has_default:
+        return lambda: smart_deepcopy(default)
+
+    # Note that unlike pydantic v1, pydantic v2 does not add a default of None when
+    # the field is Optional[something]
+    # so there is no need to handle that case here
```

Comment:
in v1 there was some logic of adding a default of `None` when its Optional. pydantic v2 doesn't do that anymore

![Uploading image.png…]()